models/Tetris.py

```diff
@@ -125,5 +148,266 @@ def rotate(self):
+
+        if not self.intersects_with_figure(curr_figure, 0, 1) and curr_figure.y + 1 < 17:
+            copied_figure = Figure(curr_figure.x, curr_figure.y + 1, curr_figure.type, curr_figure.color,
+                                    curr_figure.rotation)
+            successors.append(
+                (copied_figure, "down", self.calc_heuristic_height(copy.deepcopy(self.field), copied_figure)))
+
+        return successors
+
+    def get_a_star_successors(self, state):
+        successors = []
+        cost = 1
+
+        actions = ["down", "rotate", "left", "right"]
+        for a in actions:
+            copied_figure = Figure(state[0], state[1], self.figure.type, self.figure.color, state[2])
+            newState = None
+            if a == "right" and state[0] < 6:
+                newState = (state[0] + 1, state[1], state[2])
+            elif a == "left" and state[0] > -4:
+                newState = (state[0] - 1, state[1], state[2])
+            elif a == "down" and state[1] < 16:
+                newState = (state[0], state[1] + 1, state[2])
+            elif a == "rotate":
+                copied_figure.rotate()
+                newState = (state[0], state[1], copied_figure.rotation)
+            if newState is not None:
+                successors.append((newState, a, cost))
+
+        # return a list of successors formatted as (new figure, action, cost)
+        return successors
+
+    # GENETIC ALGORITHM:
+    # Calculate best state for given state and piece based on optimal play
+    def get_best_state(self, data):
+        for _ in range(10):
+            # drop optimal moves
+            best_move, best_score = self.get_best_move() # (self.figure.x, self.figure.rotation)
+            print("BEST MOVE: ", best_move)
+            print("BEST SCORE: ", best_score)
+            self.figure.x = best_move[0]
+            self.figure.rotation = best_move[1]
+
+            while not self.intersects():
+                self.figure.y += 1
+            self.figure.y -= 1
+            self.freeze()
+            print("FINAL STATE: ")
+            self.get_string_field()
+            time.sleep(5)
+        return self, best_score
+
+
+    # Calculate best move (piece drop) for given state
+    def get_best_move(self):
+        best_score = 999999
+        best_move = None 
+        copied_figure = self.figure
+        # calculate best move for each rotation
+        for rotate_num in rotation_map.keys():
+            # for every column drop piece and calculate score
+            for col in range(self.width - 3): # TODO: FIX RANGE? (cuts off early)
+                # simulate new states
+                copied_state = copy.deepcopy(self)
+                copied_state.figure = copied_figure
+                copied_state.figure.x = col
+                for _ in range(rotate_num):
+                    copied_state.figure.rotate()
+                print("COL", col)
+                while not copied_state.intersects():
```

Comment:
Lines 283-285 is supposed to simulate a 'drop', but it tends to intersect with existing pieces below it, which is weird

Could it be related to the column/x range stuff?